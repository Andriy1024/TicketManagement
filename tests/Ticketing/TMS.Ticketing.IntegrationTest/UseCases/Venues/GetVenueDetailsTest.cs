using FluentAssertions;

using MediatR;

using Microsoft.Extensions.DependencyInjection;

using NSubstitute;

using System.Net;

using TMS.Common.Errors;
using TMS.Test.Common;
using TMS.Ticketing.Application.Dtos;
using TMS.Ticketing.Application.Repositories;
using TMS.Ticketing.Application.Services.Payments;
using TMS.Ticketing.Application.UseCases.Venues;
using TMS.Ticketing.IntegrationTest.Common;
using TMS.Ticketing.IntegrationTest.Common.Factories;
using TMS.Ticketing.IntegrationTest.UseCases.Venues.Common;

namespace TMS.Ticketing.IntegrationTest.UseCases.Venues;

[Collection(VenuesDatabaseCollection.Name)]
public class GetVenueDetailsTest
{
    private readonly TicketingServices _services;

    private readonly static Guid DatabaseName = Guid.NewGuid();

    public GetVenueDetailsTest(MongoDbFactory mongoDb)
    {
        _services = new TicketingServices()
          .AddJsonConfig("appsettings", "appsettings.test.json")
          .AddMongoConnection(mongoDb.ConnectionString, DatabaseName.ToString())
          .BuildConfiguration()
          .AddTicketingServices()
          .OverrideService(Substitute.For<IPaymentsService>());
    }

    [Fact]
    public async Task GetVenueDetails_ByDefaultReturns_VenueDetailsDto()
    {
        // Arrange
        var venueId = Guid.NewGuid();

        await SeedDataAsync(venueId);

        VenueDetailsDto actual;

        // Act
        using (var scope = _services.BuildServicesScope())
        {
            var handler = scope.ServiceProvider.GetRequiredService<IRequestHandler<GetVenueDetails, VenueDetailsDto>>();

            actual = await handler.Handle(new GetVenueDetails(venueId), CancellationToken.None);
        }

        // Assert
        actual.Should().NotBeNull();

        actual.Should().MatchSnapshot(nameof(actual), matchOptions: o =>
        {
            o.IgnoreField($"**.{nameof(VenueOverviewDto.Id)}");
            o.IgnoreField($"**.{nameof(SectionDto.VenueId)}");
            o.IgnoreField($"**.{nameof(SectionDto.SectionId)}");
            o.IgnoreField($"**.{nameof(SeatDto.SeatId)}");

            return o;
        });
    }

    [Fact]
    public async Task GetVenueDetails_NotExistedVenue_ThrowsNotFoundError()
    {
        // Arrange
        await SeedDataAsync(Guid.NewGuid());

        Exception? actual = null;

        // Act
        using (var scope = _services.BuildServicesScope())
        {
            var handler = scope.ServiceProvider.GetRequiredService<IRequestHandler<GetVenueDetails, VenueDetailsDto>>();

            try
            {
                _ = await handler.Handle(new GetVenueDetails(Guid.NewGuid()), CancellationToken.None);
            }
            catch (Exception e)
            {
                actual = e;
            }
        }

        // Assert
        actual.Should().NotBeNull()
            .And.BeAssignableTo<ApiException>()
            .Subject.Error
            .Should().Match<ApiError>(x => x.StatusCode == HttpStatusCode.NotFound);
    }

    private async Task SeedDataAsync(Guid venueId)
    {
        using var scope = _services.BuildServicesScope();

        var repo = scope.ServiceProvider.GetRequiredService<IVenuesRepository>();

        var firstVenue = VenueTestFactory.Create(venueId: venueId, name: "Venue #3");

        await repo.AddAsync(firstVenue);
    }
}