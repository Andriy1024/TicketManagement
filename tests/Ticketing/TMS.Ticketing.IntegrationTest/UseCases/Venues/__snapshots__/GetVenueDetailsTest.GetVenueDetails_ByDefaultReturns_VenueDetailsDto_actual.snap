{
  "Sections": [
    {
      "SectionId": "163da73d-fb49-474f-b4d3-a0d4a51613db",
      "VenueId": "831d534f-4859-4af3-8cd6-8898d7b7a923",
      "Name": "Section 1",
      "Type": "Designated",
      "Seats": [
        {
          "SeatId": "f1a7a4cb-d2da-498b-87e6-25ca72fefd32",
          "SeatNumber": 1,
          "RowNumber": 1
        },
        {
          "SeatId": "061b1a51-5e76-425e-9881-ba1b90fe17ac",
          "SeatNumber": 2,
          "RowNumber": 1
        }
      ]
    }
  ],
  "Id": "831d534f-4859-4af3-8cd6-8898d7b7a923",
  "Name": "Venue #3",
  "Country": "USA",
  "City": "Denver",
  "Street": "Steet 45",
  "Details": [
    {
      "Name": "Detail 1",
      "Value": "Detail Value"
    }
  ]
}
